lib/ecto/query/planner.ex

```diff
@@ -2280,6 +2280,39 @@ defmodule Ecto.Query.Planner do
     end)
   end
 
+  defp handle_subquery_struct(select, requested_fields, ix, query) do
+    available_fields = subquery_source_fields(select)
+    requested_fields = List.wrap(requested_fields)
+
+    Enum.each(requested_fields, fn field ->
+      unless field in available_fields do
+        error!(query, "field `#{field}` in struct/2 is not available in the subquery. " <>
+                     "Subquery only returns fields: #{inspect(available_fields)}")
+      end
+    end)
+
+    schema =
+      case select do
+        {:struct, schema, _} -> schema
+        {:source, {_, schema}, _, _} when not is_nil(schema) -> schema
+        _ ->
+          error!(query, "it is not possible to return a struct subset of a subquery that does not return a schema or a struct")
+      end
+
+    types =
+      Enum.map(requested_fields, fn field ->
```

Comment:
Let's avoid double traversal of the fields. We can probably merge `subquery_type_for` and match on errors to detect the field is invalid.